packages/openai-compatible/src/openai-compatible-provider.ts

```diff
@@ -0,0 +1,163 @@
+import {
+  EmbeddingModelV1,
+  LanguageModelV1,
+  ProviderV1,
+} from '@ai-sdk/provider';
+import {
+  FetchFunction,
+  loadApiKey,
+  withoutTrailingSlash,
+} from '@ai-sdk/provider-utils';
+import { OpenAICompatibleChatLanguageModel } from './openai-compatible-chat-language-model';
+import { OpenAICompatibleChatSettings } from './openai-compatible-chat-settings';
+import { OpenAICompatibleCompletionLanguageModel } from './openai-compatible-completion-language-model';
+import { OpenAICompatibleCompletionSettings } from './openai-compatible-completion-settings';
+import { OpenAICompatibleEmbeddingSettings } from './openai-compatible-embedding-settings';
+import { OpenAICompatibleEmbeddingModel } from './openai-compatible-embedding-model';
+
+export interface OpenAICompatibleProvider<
+  L extends string = string,
+  C extends string = string,
+  E extends string = string,
+> extends ProviderV1 {
+  (modelId: L, settings?: OpenAICompatibleChatSettings): LanguageModelV1;
+
+  languageModel(
+    modelId: L,
+    settings?: OpenAICompatibleChatSettings,
+  ): LanguageModelV1;
+
+  chatModel(
+    modelId: L,
+    settings?: OpenAICompatibleChatSettings,
+  ): LanguageModelV1;
+
+  completionModel(
+    modelId: C,
+    settings?: OpenAICompatibleCompletionSettings,
+  ): LanguageModelV1;
+
+  textEmbeddingModel(
+    modelId: E,
+    settings?: OpenAICompatibleEmbeddingSettings,
+  ): EmbeddingModelV1<string>;
+}
+
+export interface OpenAICompatibleProviderSettings {
+  /**
+Base URL for the API calls.
+     */
+  baseURL?: string;
+
+  /**
+API key for authenticating requests.
+     */
+  apiKey?: string;
+
+  /**
+Custom headers to include in the requests.
+     */
+  headers?: Record<string, string>;
+
+  /**
+Custom fetch implementation. You can use it as a middleware to intercept requests,
+or to provide a custom fetch implementation for e.g. testing.
+    */
+  fetch?: FetchFunction;
+
+  /**
+The name of the environment variable from which to load the API key (if a key isn't explicitly provided).
+   */
+  apiKeyEnvVarName?: string;
+
+  /**
+Description of the API key environment variable (for use in error messages).
+   */
+  apiKeyEnvVarDescription?: string;
+
+  /**
+Provider name. Overrides the `openai` default name for 3rd party providers.
+   */
+  name?: string;
+}
+
+/**
+Create an OpenAICompatible provider instance.
+ */
+export function createOpenAICompatible<
+  L extends string,
+  C extends string,
+  E extends string,
+>(
+  options: OpenAICompatibleProviderSettings,
+): OpenAICompatibleProvider<L, C, E> {
+  // TODO(shaper):
+  // - consider throwing if baseUrl, name, sufficient api key info not available
+  // - force only 'compatible' -- look into whether we can remove some 'strict' logic/configs entirely
```

Comment:
yes - goal is that this works w/ compatible mode but is more flexible. then eventually openai provider should only be strict mode and we can simplify